RavenDB-20050: handle resharding in subscription

garayx opened the pull request:
### Issue link

https://issues.hibernatingrhinos.com/issue/RavenDB-20050

### Additional description

- Halt subscription processing on shard if we found document from different shard or document that is under active migration additionally.
- Track subscription process on orchestrator so we can filter documents that were already sent

### Type of change

- Bug fix

### How risky is the change?

- Not relevant

### Backward compatibility

- Not relevant

### Is it platform specific issue?

- No

### Documentation update

- No documentation update is needed 

### Testing by Contributor

- Tests have been added that prove the fix is effective or that the feature works
 - Internal classes added to the test class (e.g. entity or index definition classes) have the lowest possible access modifier (preferable `private`) 
- It has been verified by manual testing

### Testing by RavenDB QA team

- No special testing by RavenDB QA team is needed

### Is there any existing behavior change of other features due to this change?

- No

### UI work

- No UI work is needed

garayx committed:
RavenDB-20050: handle resharding in subscription

karmeli87 commented on src/Raven.Server/Documents/DocumentsStorage.cs:
```diff
@@ -1920,7 +1920,6 @@ public void EnsureLastEtagIsPersisted(DocumentsOperationContext context, long do
                     ChangeVectorUtils.ThrowConflictingEtag(lowerId.ToString(), docChangeVector, newEtag, Environment.Base64Id, DocumentDatabase.ServerStore.NodeTag);
 
                 context.LastDatabaseChangeVector ??= GetDatabaseChangeVector(context);
-                context.LastDatabaseChangeVector = ChangeVector.Merge(context.LastDatabaseChangeVector, context.GetChangeVector(mergedChangeVector), context);
```
please make a comment here with this issue
https://issues.hibernatingrhinos.com/issue/RavenDB-20732/Deleting-a-document-that-was-resharded-cause-a-change-vector-leak

karmeli87 commented on src/Raven.Server/Documents/Subscriptions/SubscriptionConnectionBase.cs:
```diff
@@ -160,9 +162,8 @@ private static string ExtractDatabaseNameForLogging(TcpConnectionOptions tcpConn
                                 var sendingCurrentBatchStopwatch = Stopwatch.StartNew();
 
-                                var anyDocumentsSentInCurrentIteration =
+                                (bool anyDocumentsSentInCurrentIteration, bool isActiveMigration) = 
```
`isActiveMigration` shouldn't be here

karmeli87 commented on src/Raven.Server/Documents/Subscriptions/SubscriptionConnectionBase.cs:
```diff
@@ -178,13 +179,27 @@ private static string ExtractDatabaseNameForLogging(TcpConnectionOptions tcpConn
 
-                                    if (FoundAboutMoreDocs())
-                                        state.NotifyHasMoreDocs();
+                                    if (isActiveMigration)
+                                    {
+                                        Debug.Assert(_subscriptions is ShardSubscriptionStorage);
+                                        if (await WaitForDocsMigrationAsync(state, replyFromClientTask))
+                                        {
+                                            // we waited for Migration and try to pull the docs again
+                                            continue;
+                                        }
+
+                                        // replyFromClientTask is completed (it will be awaited in WaitForClientAck)
+                                    }
```
This shouldn't be in the base class

ppekrol commented on src/Raven.Server/Documents/Sharding/Subscriptions/OrchestratedSubscriptionConnection.cs:
```diff
@@ -26,6 +28,7 @@ public class OrchestratedSubscriptionConnection : SubscriptionConnectionBase<Orc
         private readonly ShardedDatabaseContext _databaseContext;
         private OrchestratedSubscriptionProcessor _processor;
         private readonly IDisposable _tokenRegisterDisposable;
+        private readonly HashSet<string> _dbIdToRemove;
```
_dbIdsToRemove

ppekrol commented on src/Raven.Server/Documents/Sharding/Subscriptions/OrchestratedSubscriptionConnection.cs:
```diff
@@ -35,6 +38,7 @@ public class OrchestratedSubscriptionConnection : SubscriptionConnectionBase<Orc
         {
             _databaseContext = tcpConnection.DatabaseContext;
             _tokenRegisterDisposable = CancellationTokenSource.Token.Register(() => _processor?.CurrentBatch?.SetCancel());
+            _dbIdToRemove = new HashSet<string>() { _databaseContext.DatabaseRecord.Sharding.DatabaseId };
```
sanity-check: no need for case insensitive comparer here?

garayx replied:
added

ppekrol commented on src/Raven.Server/Documents/Sharding/Subscriptions/OrchestratedSubscriptionConnection.cs:
```diff
@@ -76,19 +83,25 @@ protected override void GatherIncludesForDocument(OrchestratorIncludesCommandImp
         {
             // nothing was sent to the client, but we need to let the shard know he can continue
-            await NotifyShardAboutBatchCompletion();
+            await NotifyShardAboutBatchCompletion(null);
```
in cases like that (passing null or bool), it is better to use named argument, so it is easier to understand what are you passing there actually

ppekrol commented on src/Raven.Server/Documents/Sharding/Subscriptions/OrchestratedSubscriptionConnection.cs:
```diff
@@ -76,19 +83,25 @@ protected override void GatherIncludesForDocument(OrchestratorIncludesCommandImp
+            return Task.FromResult(true);
+        }
+
+        private async Task NotifyShardAboutBatchCompletion(string orchestratorCv)
```
Async :)

ppekrol commented on src/Raven.Server/Documents/Sharding/Subscriptions/ShardedDocumentsDatabaseSubscriptionProcessor.cs:
```diff
@@ -47,32 +55,49 @@ protected override bool ShouldSend(Document item, out string reason, out Excepti
-            var bucket = ShardHelper.GetBucketFor(_sharding, _allocator, item.Id);
-           
-            if (_sharding.BucketMigrations.TryGetValue(bucket, out var migration))
+            IsActiveMigration = isActiveMigration;
```
I think it is better to pack all `out` arguments into some struct, also this is per item, so should be there?

garayx replied:
done

ppekrol commented on src/Raven.Server/Documents/Sharding/Subscriptions/ShardedDocumentsDatabaseSubscriptionProcessor.cs:
```diff
@@ -109,6 +147,14 @@ public override async Task<long> RecordBatch(string lastChangeVectorSentInThisBa
         return result.Index;
     }
 
+    public override async Task AcknowledgeBatch(long batchId, string changeVector)
```
Asyn

ppekrol commented on src/Raven.Server/Documents/Subscriptions/SubscriptionConnectionBase.cs:
```diff
@@ -51,6 +52,7 @@ public abstract class SubscriptionConnectionBase<TIncludesCommand> : ISubscripti
         internal SubscriptionWorkerOptions _options;
         internal readonly Logger _logger;
 
+        public string LastSentChangeVectorInThisConnection;
```
you are passing a lot of state via instance fields, please check if this is really necessary

garayx replied:
I just moved it from `SubscriptionConnection` to base class (so I can use it in orchestrator).
In general AFAIK we have this field, since in concurrent subscription it saves the cv for particular connection, and the field in state save the merged one.

ppekrol commented on src/Raven.Server/Documents/Subscriptions/SubscriptionConnectionsState.cs:
```diff
@@ -110,12 +110,20 @@ public HashSet<long> GetActiveBatches()
         }
 
-        public Task AcknowledgeBatch(SubscriptionConnection connection, long batchId, List<DocumentRecord> addDocumentsToResend)
+        public async Task AcknowledgeShardingBatch(string changeVector, string cvForOrchestrator, long batchId, List<DocumentRecord> addDocumentsToResend)
```
Async :)

ppekrol commented on src/Raven.Server/Documents/Subscriptions/SubscriptionConnectionsState.cs:
```diff
@@ -110,12 +110,20 @@ public HashSet<long> GetActiveBatches()
+            await WaitForIndexNotificationAsync(etag);
+        }
+
+        public async Task AcknowledgeBatch(string changeVector, long batchId, List<DocumentRecord> addDocumentsToResend)
```
Async :)

garayx committed:
RavenDB-20050: refactoring

karmeli87 commented on src/Raven.Server/Documents/Subscriptions/SubscriptionConnectionBase.cs:
```diff
@@ -209,28 +224,65 @@ private static string ExtractDatabaseNameForLogging(TcpConnectionOptions tcpConn
             {
                 AddToStatusDescription(CreateStatusMessage(ConnectionStatus.Info, "Start trying to send docs to client"));
                 bool anyDocumentsSentInCurrentIteration = false;
+                bool isActiveMigration = false;
```
can we remove this from the base class?

garayx replied:
yes

karmeli87 commented on src/Raven.Server/Documents/Subscriptions/SubscriptionConnectionBase.cs:
```diff
@@ -254,35 +306,66 @@ await using (var writer = new AsyncBlittableJsonTextWriter(context, _buffer))
+                                }
 
-                                    continue;
+                                if (result.IsActiveMigration)
```
instead of the all of the `if IsActiveMigration` would be simpler to throw `DocumentUnderActiveMigrationException` in `GetBatch`?

garayx replied:
no since, I still want to send the items that were fetched before the item that is under active migration

ppekrol replied:
This can be done by introducing a method `HandleBatchItem` and overriding for sharding

garayx replied:
done

karmeli87 commented on src/Raven.Server/Documents/Subscriptions/SubscriptionConnectionBase.cs:
```diff
@@ -254,35 +306,66 @@ await using (var writer = new AsyncBlittableJsonTextWriter(context, _buffer))
+                                    return BatchStatus.ActiveMigration;
+                                }
+
+                                // we already pulled some docs, we will send the docs in CurrentBatch and then will wait for migration to complete
```
Is this an optimization? if so, maybe simpler to break `GetBatch` earlier ?

garayx replied:
yes I want to send all items that were fetched prior to item that is active migration

ppekrol replied:
all of those I think can be relatively easly handled by introducing methods and just overriding them?

karmeli87 replied:
if we drop this optimization, would that simplify the code?

garayx replied:
> all of those I think can be relatively easly handled by introducing methods and just overriding them?

yes

karmeli87 commented on src/Raven.Server/Documents/Subscriptions/SubscriptionProcessor/AbstractSubscriptionProcessorBase.cs:
```diff
@@ -0,0 +1,17 @@
+
+namespace Raven.Server.Documents.Subscriptions.SubscriptionProcessor;
+
+public abstract class AbstractSubscriptionProcessorBase : IDisposable
```
why do we need this?

garayx replied:
so I don't have to use `AbstractSubscriptionProcessor<T>` to use `BatchItem`

karmeli87 commented on src/Raven.Server/Documents/Subscriptions/SubscriptionProcessor/DatabaseSubscriptionProcessor.cs:
```diff
@@ -40,18 +41,29 @@ public override IDisposable InitializeForNewBatch(ClusterOperationContext cluste
+        protected BatchItem GetBatchItem(T item)
         {
-            if (ShouldSend(item, out var reason, out var exception, out var result))
+            if (ShouldSend(item, out var reason, out var exception, out var result, out var isActiveMigration))
```
remove `isActiveMigration` from here

karmeli87 replied:
maybe make an `out BatchItem batchItem`

garayx replied:
> remove `isActiveMigration` from here

why ?

karmeli87 commented on src/Raven.Server/Documents/Subscriptions/SubscriptionProcessor/DocumentsDatabaseSubscriptionProcessor.cs:
```diff
@@ -111,15 +112,16 @@ protected override SubscriptionFetcher<Document> CreateFetcher()
         }
 
-        protected override bool ShouldSend(Document item, out string reason, out Exception exception, out Document result)
+        protected override bool ShouldSend(Document item, out string reason, out Exception exception, out Document result, out bool isActiveMigration)
```
remove `isActiveMigration`

karmeli87 commented on src/Raven.Server/Documents/Subscriptions/SubscriptionProcessor/DocumentsDatabaseSubscriptionProcessor.cs:
```diff
@@ -221,17 +215,28 @@ protected virtual bool ShouldFetchFromResend(DocumentsOperationContext context,
                     return fetch;
 
                 case ConflictStatus.AlreadyMerged:
+                    if (CheckIfNewerInResendList(context, item.Document.Id, item.Document.ChangeVector, currentChangeVector))
```
this is something that is relevant for 5.x as well?

garayx replied:
I cannot think of such situation in concurrent subscription

karmeli87 commented on src/Raven.Server/Documents/TcpHandlers/SubscriptionConnection.cs:
```diff
@@ -453,6 +454,11 @@ protected override void GatherIncludesForDocument(DatabaseIncludesCommandImpl in
             includeDocuments?.GatherIncludesForDocument(document);
         }
 
+        protected override Task<bool> WaitForDocsMigrationAsync(AbstractSubscriptionConnectionsState state, Task pendingReply)
```
this shouldn't be here

garayx replied:
done

karmeli87 commented on src/Raven.Server/Documents/TcpHandlers/SubscriptionConnectionForShard.cs:
```diff
@@ -82,12 +92,25 @@ public override AbstractSubscriptionProcessor<DatabaseIncludesCommandImpl> Creat
+            {
+                _logger.Info($"Got '{nameof(DocumentUnderActiveMigrationException)}' on shard '{ShardName}' will roll back the change vector from '{LastSentChangeVectorInThisConnection}' to '{cvBeforeRecordBatch}', and wait for migration to complete.", e);
+            }
+            LastSentChangeVectorInThisConnection = cvBeforeRecordBatch;
```
to simplify the flow... consider keep throwing this instead

karmeli87 commented on src/Raven.Client/Documents/Subscriptions/SubscriptionShardingState.cs:
```diff
@@ -9,6 +9,7 @@ public class SubscriptionShardingState : IDynamicJson
     public Dictionary<string, string> ChangeVectorForNextBatchStartingPointPerShard { get; set; }
     public Dictionary<string, string> NodeTagPerShard { get; set; }
     public Dictionary<int, string> ProcessedChangeVectorPerBucket { get; set; }
+    public string ChangeVectorForNextBatchStartingPointForOrchestrator { get; set; }
```
we have `LastChangeVectorSent` vs `ChangeVectorForNextBatchStartingPointForOrchestrator` is there any difference for sharded database?

garayx replied:
`ChangeVectorForNextBatchStartingPointForOrchestrator` used to persist the cv to storage, `LastChangeVectorSent` is used in memory in `SubscriptionConnection`. I want to have a meaningful name since this state is used by orchestrator and shards and regular subscription

ppekrol commented on src/Raven.Server/Documents/Subscriptions/SubscriptionConnectionBase.cs:
```diff
@@ -185,16 +176,40 @@ private static string ExtractDatabaseNameForLogging(TcpConnectionOptions tcpConn
+
+                                    break;
+
+                                case BatchStatus.ActiveMigration:
```
you can easily solve this flaw by introducing a method:

`protected virtual bool HandleBatchStatus(...)`

and adding in this base class 2 common cases and in sharded one overriding it and doing:

```
if (status == ActiveMigration)
{
    DoWork();
    return true;
}

return base.HandleBatchStatus();

```

garayx replied:
done

ppekrol commented on src/Raven.Server/Documents/Sharding/Subscriptions/OrchestratedSubscriptionConnection.cs:
```diff
@@ -63,8 +69,11 @@ protected override StatusMessageDetails GetStatusMessageDetails()
         protected override async Task OnClientAckAsync(string clientReplyChangeVector)
         {
-            await NotifyShardAboutBatchCompletion();
+            var orchestratorCv = ChangeVectorUtils.MergeVectors(_state.LastChangeVectorSent, LastSentChangeVectorInThisConnection);
```
I think you need to merge the Order part of CV only

garayx replied:
we want to merge the Order part, and its is done in `TryRecordBatchAndUpdateStatusAsync` so `LastSentChangeVectorInThisConnection` is single cv here

ppekrol commented on src/Raven.Server/Documents/Sharding/Subscriptions/OrchestratedSubscriptionConnection.cs:
```diff
@@ -73,22 +82,23 @@ protected override void GatherIncludesForDocument(OrchestratorIncludesCommandImp
             {
                 // let sharded subscription worker know that we sent the batch to the client and received an ack request from it
                 batch.SendBatchToClientTcs.TrySetResult();
+                batch.LastSentChangeVectorInBatch = orchestratorCv; //pass the orchestrator merged change vector to shard, so it will ack it
```
this is fine to set null here in WaitForChangedDocsAsync? Can you explain this case?

garayx replied:
it seems like a precaution for empty batch received on orchestrator (so we will advance the shard),
according to the code if its null then in `AcknowledgeSubscriptionBatchCommand` the `currentState.ShardingState.ChangeVectorForNextBatchStartingPointForOrchestrator` will stay the same which is fine (since this was empty batch)

ppekrol commented on src/Raven.Server/Documents/Sharding/Subscriptions/OrchestratedSubscriptionConnection.cs:
```diff
@@ -109,10 +119,11 @@ protected override bool FoundAboutMoreDocs()
         }
 
-        public override IDisposable MarkInUse()
+        public override DisposeOnce<SingleAttempt> MarkInUse()
```
hmm?

garayx replied:
after refactoring I may have double dispose, so had to change this :)

ppekrol replied:
Let's talk about that tomorrow

ppekrol commented on src/Raven.Server/Documents/Sharding/Subscriptions/OrchestratedSubscriptionConnection.cs:
```diff
@@ -125,13 +136,18 @@ protected override void RaiseNotificationForBatchEnd(string name, SubscriptionBa
+            var vector = context.GetChangeVector(result.LastChangeVectorSentInThisBatch);
+            vector.TryRemoveIds(_dbIdsToRemove, context, out vector);
+
+            LastSentChangeVectorInThisConnection = vector.Order;
```
Shouldn't we merge this with prev value?

garayx replied:
we merge & persist this on `OnClientAckAsync` method
basically we apply the legacy subsciption logic for orchestrator: 
1. save cv in memory when fetching, 
3. persist it after receiving ack from client

ppekrol commented on src/Raven.Server/Documents/Sharding/Subscriptions/OrchestratedSubscriptionConnection.cs:
```diff
@@ -125,13 +136,18 @@ protected override void RaiseNotificationForBatchEnd(string name, SubscriptionBa
-            DevelopmentHelper.ShardingToDo(DevelopmentHelper.TeamMember.Karmel, DevelopmentHelper.Severity.Normal, "RavenDB-19085 All tests pass but wonder if that is correct?");
-            return sentDocument.ChangeVector;
-        }
+            var vector = context.GetChangeVector(result.LastChangeVectorSentInThisBatch);
```
in one case you are setting null here, can this occur?

garayx replied:
yes `result.LastChangeVectorSentInThisBatch` can be null, this is fine, the batch status will be `EmptyBatch`

ppekrol commented on src/Raven.Server/Documents/Subscriptions/SubscriptionConnectionBase.cs:
```diff
@@ -209,38 +177,107 @@ private static string ExtractDatabaseNameForLogging(TcpConnectionOptions tcpConn
+                    break;
+
+                case BatchStatus.DocumentsSent:
+                    markInUse.Dispose();
```
why we are disposing this here? the outer using is not enough?

garayx replied:
in previous code the `WaitForClientAck` happens after markInUse is disposed:
https://github.com/ravendb/ravendb/blob/3b0f69bd2a70dc5a831eb5a49467e77087b43217/src/Raven.Server/Documents/Subscriptions/SubscriptionConnectionBase.cs#L193

ppekrol commented on src/Raven.Server/Documents/Subscriptions/SubscriptionConnectionsState.cs:
```diff
@@ -110,12 +110,20 @@ public HashSet<long> GetActiveBatches()
         }
 
-        public Task AcknowledgeBatch(SubscriptionConnection connection, long batchId, List<DocumentRecord> addDocumentsToResend)
+        public async Task AcknowledgeShardingBatchAsync(string changeVector, string cvForOrchestrator, long batchId, List<DocumentRecord> addDocumentsToResend)
```
this should not be in this class?

ppekrol replied:
shouldn't you just override the 'AcknowledgeBatchAsync' in one of the classes that inherit from this?

garayx replied:
I had to pass another parameter to the method: `string cvForOrchestrator` so I cannot simply override `AcknowledgeBatchAsync`

ppekrol commented on src/Raven.Server/Documents/TcpHandlers/SubscriptionConnection.cs:
```diff
@@ -383,7 +383,7 @@ protected override bool FoundAboutMoreDocs()
         }
 
-        public override IDisposable MarkInUse() => _database.DatabaseInUse(skipUsagesCount: false);
+        public override DisposeOnce<SingleAttempt> MarkInUse() => new DisposeOnce<SingleAttempt>(() => _database.DatabaseInUse(skipUsagesCount: false));
```
this looks weird

garayx replied:
need to make sure we dispose once now

ppekrol commented on src/Raven.Server/ServerWide/Commands/Subscriptions/AcknowledgeSubscriptionBatchCommand.cs:
```diff
@@ -21,7 +20,10 @@ namespace Raven.Server.ServerWide.Commands.Subscriptions
+
+        // in regular subscription LastKnownSubscriptionChangeVector used only for backward compatibility
+        // in sharing subscription LastKnownSubscriptionChangeVector used to save orchestrator cv
+        public string LastKnownSubscriptionChangeVector; 
```
what we will have here then for orchestrator vs ChangeVector field?

garayx replied:
for shard it will be:
`ChangeVector` = local change vector for shard
 `LastKnownSubscriptionChangeVector` = merged change vector of the orchestrator
 
 both will be persisted in this command in to the `ShardingState` class

ppekrol reviewed (commented):
Will continue tomorrow...

ppekrol committed:
RavenDB-20050
- refactored ShouldSend method
- renamed HandleBatchStatus => HandleBatchStatusAsync

ppekrol committed:
RavenDB-20050
- renamed GetBatch => GetBatchAsync
- removed unnecessary async/await in methods
- removed AcknowledgeShardingBatchAsync

ppekrol committed:
RavenDB-20050 renames

ppekrol committed:
RavenDB-20050 CanContinueBatch does not need to be async

ppekrol committed:
RavenDB-20050 minor

ppekrol committed:
RavenDB-20050 do not use DisposeOnce

ppekrol committed:
RavenDB-20050 abstracted HandleBatchItem and GetBatchItem

garayx commented on src/Raven.Server/Documents/Subscriptions/Processor/DatabaseSubscriptionProcessor.cs:
```diff
@@ -1,24 +1,22 @@
+namespace Raven.Server.Documents.Subscriptions.Processor
 {
-    public abstract class DatabaseSubscriptionProcessor<T> : DatabaseSubscriptionProcessor
+    public abstract class DatabaseSubscriptionProcessor<T> : DatabaseSubscriptionProcessor2<T>
```
```suggestion
    public abstract class DatabaseSubscriptionProcessor<T> : DatabaseSubscriptionProcessorBase<T>
```

garayx commented on src/Raven.Server/Documents/Subscriptions/Processor/DocumentsDatabaseSubscriptionProcessor.cs:
```diff
@@ -24,26 +24,28 @@ public class DocumentsDatabaseSubscriptionProcessor : DatabaseSubscriptionProces
+                if (CanContinueBatch(batchItem, size, result.CurrentBatch.Count, sendingCurrentBatchStopwatch) == false)
                     break;
+
+                await SendHeartbeatIfNeededAsync(sendingCurrentBatchStopwatch);
```
we need to send Heartbeat also when `CanContinueBatch  == false`

garayx committed:
RavenDB-20050: use SubscriptionBatchItemStatus in HandleBatchItem and dispose skipped BatchItem

garayx committed:
RavenDB-20050: handle Exception SubscriptionBatchItemStatus
